lib/datasource.js

```diff
@@ -2115,26 +2157,125 @@ DataSource.prototype.transaction = function() {
 
   transaction.isTransaction = true;
   transaction.origin = dataSource;
-  transaction.name = dataSource.name;
-  transaction.settings = dataSource.settings;
   transaction.connected = false;
   transaction.connecting = false;
-  transaction.connector = dataSource.connector.transaction();
 
   // create blank models pool
-  transaction.modelBuilder = new ModelBuilder();
-  transaction.models = transaction.modelBuilder.models;
-  transaction.definitions = transaction.modelBuilder.definitions;
+  var modelBuilder = new ModelBuilder();
+  var slaveModels = modelBuilder.models;
+  transaction.modelBuilder = modelBuilder;
+  transaction.models = slaveModels;
+  transaction.definitions = modelBuilder.definitions;
+
+  // For performance reasons, use a getter per model and only call copyModel()
+  // for the models that are actually used. These getters are then replaced
+  // with the actual values on first use.
+  var masterModels = dataSource.modelBuilder.models;
+  Object.keys(masterModels).forEach(function(name) {
+    Object.defineProperty(slaveModels, name, {
+      enumerable: true,
+      configurable: true,
+      get: function() {
+        // Delete getter so copyModel() can redefine slaveModels[name].
+        delete slaveModels[name];
+        return dataSource.copyModel.call(transaction, masterModels[name]);
+      },
+    });
+  });
+
+  function handleExecute() {
+    if (execute) {
+      cb = cb || utils.createPromiseCallback();
+      var done = function(err) {
+        if (err) {
+          transaction.rollback(function(error) {
+            cb(err || error);
+          });
+        } else {
+          transaction.commit(cb);
+        }
+        // Make sure cb() isn't called twice, e.g. if `execute` returns a
+        // thenable object and also calls the passed `cb` function.
+        done = function() {};
+      };
+      try {
+        var result = execute(slaveModels, done);
+        if (result && typeof result.then === 'function') {
+          result.then(function() { done(); }, done);
+        } else if (execute.length < 2) {
+          // execute didn't return a thenable and doesn't receive a callback.
+          // It must executed synchronously, so call done() now.
```

Comment:
Minor nitpick:
// It must've executed synchronously, so call done now.

Reply:
Addressed!